sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,276 @@
+all non-`SYCL_EXTERNAL` functions are internalized to avoid multiple definition
+errors during runtime linking.
+Device images with `SYCL_EXTERNAL` functions will also have a list of names
+of exported functions.
+
+**NOTE**: If device code split is enabled, it seems reasonable to perform
+exports arrangement before device code split procedure.
+
+In orger to collect information about imported symbols `sycl-post-link` looks
+through LLVM IR and for each declared but not defined symbol records its name,
+except the following cases:
+- Declarations with `__` prefix in demangled name are not recorded as imported
+  functions
+  - Declarations with `__spirv_*` prefix should not be recorded as dependencies
+  since they represent SPIR-V operations and will be transformed to SPIR-V
+  instructions during LLVM->SPIR-V translation.
+- Based on some attributes which could be defined later
+  - This is needed to have possibility to call device-specific builtins not
+    starting with `__` by forward-declaring them in DPC++ code
+
+**NOTE**: If device code split is enabled, imports collection is performed after
+split and it is performed on splitted images.
+
+All collected information is attached to a device image via properties
+mechanism.
+
+Each device image is supplied with an array of property sets:
+```
+struct pi_device_binary_struct {
+...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+
+```
+Each property set is represent by the following struct:
+```
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`ImportedSymbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `ExportedSymbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
+images required to execute a kernel based on the list of exports/imports and
+links them together using PI API.
+
+Given that all exports will be arranged to a separate device images without
+kernels it is reasonable to store device images with exports in a separate data
+structure.
+
```

Comment:
Please add some words about device code caching in presence of dynamic linkage. Will it work?
Runtime-only caching or persistent (on-disk) too?

Reply:
- Does it make sense to have an API to force loading device code similar to `dlopen` - something like `sycl::INTEL::devdlopen("foo.spv")` to make foo.spv part of dynamic symbol resolution?
- AFAIU, device code modules loaded via other mechanisms - L0, OpenCL - won't be used for dynamic resolution - correct?

Reply:
> Does it make sense to have an API to force loading device code similar to `dlopen` - something like `sycl::INTEL::devdlopen("foo.spv")` to make foo.spv part of dynamic symbol resolution?

If someone finds this useful, we could do that, I don't see any obstacles here

> AFAIU, device code modules loaded via other mechanisms - L0, OpenCL - won't be used for dynamic resolution - correct?

At least this is my vision of this feature as I stated it here: https://github.com/intel/llvm/pull/3210#discussion_r575388163

Reply:
OK, thanks. What about caching?

Reply:
Described caching. I beleive in-memory cache should work without out of the box, however we may need some changes in device images identifying mechanism for persistent cache.

Reply:
It looks like the plan is to JIT compile the `SYCL_EXTERNAL` functions into native code and then cache the results **before** they are linked into the calling kernel.  If we find a kernel that calls one of these functions, the calling kernel will be compiled to native code, and then we will link the two native pieces together (the kernel and the `SYCL_EXTERNAL` function).  Is that correct?

We considered a similar design for the "device library" code, but we rejected it.  The problem is that you miss opportunity for cross-function optimization in this case, which is particularly important for GPU code.  It's much better to link the two SPIR-V modules together first, and then compile the whole thing into native code.  This way, the JIT compiler can optimize across the call site to the `SYCL_EXTERNAL` function.

If you follow this approach, you wouldn't need to cache native code for any of the `SYCL_EXTERNAL` functions.  Instead, you would only cache native code for the device images that contain complete kernels.

Reply:
I was following the logic of OpenCL `clCompileProgram` and `clLinkProgram` assuming that available PI API will work in the same way. `clLinkProgram` accepts only compiled programs, so the plan was to use `clCompileProgram` on library device image and store it in cache. Probably I was under impression that `clCompileProgram` API is invocation of frontend, but since our main format is SPIR-V - then frontend was already ran and most likely `clCompileProgram` will be a noop.
Is there any API to link device images before they compiled (or is there any API to link SPIR-V device images)?

Reply:
> Probably I was under impression that clCompileProgram API is invocation of frontend, but since our main format is SPIR-V - then frontend was already ran and most likely clCompileProgram will be a noop.

Correct, `clCompileProgram()` is a noop in this case and it's `clLinkProgram()` that really does the compilation from SPIR-V to native.

Reply:
Note: it might not be a pure noop, for example `clCompileProgram` will perform SPIR-V specialization: https://github.com/KhronosGroup/OpenCL-Docs/issues/553